[analyzer] Extend EnumCastOutOfRange diagnostics

gamesh411 opened the pull request:
EnumCastOutOfRange checker now reports the name of the enum in the warning message. Additionally, a note-tag is placed to highlight the location of the declaration.

Commit:
[analyzer] Extend EnumCastOutOfRange diagnostics

EnumCastOutOfRange checker now reports the name of the enum in the
warning message. Additionally, a note-tag is placed to highlight the
location of the declaration.

Commit:
Track expression value of the CastExpr

Xazax-hun commented:
Overall looks good, but I think there are no tests that would showcase what trackexpression is actually doing.

Commit:
add trackexpression test
track subexpression instead of the whole cast

gamesh411 commented:
@Xazax-hun Thanks for the review!
I have modified the tracking a bit, because seeing the other usecases, and also just thinking through, the *subexpression* of the cast is what we are interested in upstream in the bugpath.
I have added a very minimal test as well, and rebased.

Commit:
Add test cases for anonymous enums
